racket/src/io/time/main.rkt

```diff
@@ -0,0 +1,37 @@
+#lang racket/base
+
+(provide seconds->date)
+
+(require racket/fixnum
+         (only-in '#%kernel [date*? kernel:date*?])
+         "../common/check.rkt"
+         "../host/rktio.rkt"
+         "../string/main.rkt"
+         "../error/main.rkt")
+
+(define rktio_seconds_to_date-error-kind
+  (vector-immutable RKTIO_ERROR_KIND_RACKET
+                    RKTIO_ERROR_TIME_OUT_OF_RANGE ))
+
+(define/who seconds->date
+  (case-lambda
+   [(s) (seconds->date s #t)]
+   [(s local?)
+    (check who real? s)
+    (let* ([s (inexact->exact s)]
+           [si (floor s)]
+           [get-gmt (if local? 0 1)]
+           [nsecs (floor (* (- s si) 1000000000))]
+           [dt (rktio_seconds_to_date* rktio si nsecs get-gmt)])
+      (cond
+        [(kernel:date*? dt)
+         dt]
+        [(equal? dt rktio_seconds_to_date-error-kind)
+         (raise-arguments-error who "integer is out-of-range"
+                                "integer" si)]
+        [else
+         (error who "conversion error\n  error: ~a; ~a"
+                (vector-ref dt 1)
+                (bytes->string/utf-8
+                 (rktio_to_bytes
+                  (rktio_get_last_error_string rktio))))]))]))
```

Comment:
Don't use `rktio_get_last_error_string`, because some other Racket thread (in the same place) may have run between the call to `rktio_seconds_to_date*` and trying to get an error. That's why all the `rktio_...` wrappers return a value for an error, where the vector is gathered as part of the atomic region. Use `raise-racket-error` from "../host/error.rkt" to turn an error vector into an exception.